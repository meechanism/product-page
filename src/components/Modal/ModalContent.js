import React from 'react';
import ReactDOM from 'react-dom';
import FocusTrap from 'focus-trap-react';
import styled from 'styled-components';

import Colors from '../../styles/colors';
import MediaQuery from '../../styles/media-queries';

const ModalWrapper = styled.aside.attrs({
  tag: 'aside',
  role: 'dialog',
  tabIndex: -1,
  'aria-modal': true
})`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10;
  transform: translateZ(0);
  background-color: rgba(0, 0, 0, 0.5);
`;

const CloseButton = styled.button.attrs((props) => ({
  'aria-label': 'Close Modal',
  'aria-labelledby': 'close-modal'
}))`
  position: absolute;
  top: 0;
  right: 0;
  padding: 0.5em;
  line-height: 1;
  background: ${Colors.trueWhite};
  border: 0;
  box-shadow: 0;
  cursor: pointer;

  &:hover {
    svg {
      stroke: ${Colors.primary};
    }
  }
`;

const ModalArea = styled.div`
  background-color: ${Colors.trueWhite};
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.1);
  border-radius: 2px;
  position: fixed;
  width: 100%;
  overflow-y: auto;
  overflow-x: hidden;
  -webkit-overflow-scrolling: touch;

  left: 50%;
  top: 50%;
  height: auto;
  transform: translate(-50%, -50%);
  max-width: 30em;
  max-height: calc(100% - 1em);

  ${MediaQuery.small`
    width: 100%;
  `};
`;

const CloseIcon = styled.svg`
  width: 25px;
  height: 25px;
  fill: transparent;
  stroke: black;
  stroke-linecap: round;
  stroke-width: 2;
`;

const CloseVisual = styled.span`
  border: 0 !important;
  clip: rect(0 0 0 0) !important;
  height: 1px !important;
  margin: -1px !important;
  overflow: hidden !important;
  padding: 0 !important;
  position: absolute !important;
  width: 1px !important;
  white-space: nowrap !important;
`;

const ModalContent = ({
  buttonRef,
  onClickOutside,
  onKeyDown,
  modalRef,
  closeModal,
  content
}) => {
  return ReactDOM.createPortal(
    <FocusTrap>
      <ModalWrapper onClick={onClickOutside} onKeyDown={onKeyDown}>
        <ModalArea ref={modalRef}>
          <CloseButton ref={buttonRef} onClick={closeModal}>
            <CloseVisual id="close-modal">Close</CloseVisual>
            <CloseIcon viewBox="0 0 40 40">
              <path d="M 10,10 L 30,30 M 30,10 L 10,30" />
            </CloseIcon>
          </CloseButton>
          <div>{content}</div>
        </ModalArea>
      </ModalWrapper>
    </FocusTrap>,
    document.body
  );
};

export default ModalContent;
