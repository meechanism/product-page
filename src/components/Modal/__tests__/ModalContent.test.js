import React from 'react';
import FocusTrap from 'focus-trap-react';

import ModalContent, {
  ModalWrapper,
  CloseButton,
  ModalArea,
  CloseIcon,
  CloseVisual
} from '../ModalContent';

describe('ModalContent component', () => {
  it('should render', () => {
    const output = shallow(<ModalContent />);
    expect(shallowToJson(output)).toMatchSnapshot();
    expect(output.find(FocusTrap).exists()).toBeTruthy();
    expect(output.find(ModalWrapper).exists()).toBeTruthy();
    expect(output.find(CloseButton).exists()).toBeTruthy();
    expect(output.find(ModalArea).exists()).toBeTruthy();
    expect(output.find(CloseIcon).exists()).toBeTruthy();
    expect(output.find(CloseVisual).exists()).toBeTruthy();
  });

  it('should trigger callback when close button clicked', () => {
    const mockFn = jest.fn();
    const output = shallow(<ModalContent closeModal={mockFn} />);
    const btn = output.find(CloseButton);
    btn.simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });

  it('should delegate provided props to children components', () => {
    const mockFn = jest.fn();
    const onClickOutside = jest.fn();
    const onKeyDown = jest.fn();
    const closeModal = jest.fn();

    const output = mount(
      <ModalContent
        closeModal={mockFn}
        onClickOutside={onClickOutside}
        onKeyDown={onKeyDown}
      />
    );
    const modalWrapper = output.find(ModalWrapper);
    expect(modalWrapper.props().onClick).toBe(onClickOutside);
    expect(modalWrapper.props().onKeyDown).toBe(onKeyDown);

    const closeBtn = output.find(CloseButton);
    expect(closeBtn.props().onClick).toBe(mockFn);
  });
});
