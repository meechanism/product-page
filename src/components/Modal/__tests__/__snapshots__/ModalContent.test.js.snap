// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContent component should render 1`] = `
<Portal
  containerInfo={<body />}
>
  <FocusTrap
    _createFocusTrap={[Function]}
    active={true}
    focusTrapOptions={Object {}}
    paused={false}
  >
    <styled.aside>
      <styled.div>
        <styled.button>
          <styled.span
            id="close-modal"
          >
            Close
          </styled.span>
          <styled.svg
            viewBox="0 0 40 40"
          >
            <path
              d="M 10,10 L 30,30 M 30,10 L 10,30"
            />
          </styled.svg>
        </styled.button>
        <div />
      </styled.div>
    </styled.aside>
  </FocusTrap>
</Portal>
`;
